core/src/HIP/Kokkos_HIP_ReduceScan.hpp

```diff
@@ -463,69 +449,88 @@ __device__ bool hip_single_inter_block_reduce_scan_impl(
 
     {
       void* const shared_ptr = shared_data + word_count.value * threadIdx.y;
-      /* reference_type shared_value = */ ValueInit::init(functor, shared_ptr);
+      /* reference_type shared_value = */ functor.init(
+          static_cast<pointer_type>(shared_ptr));
 
       for (size_type i = b; i < e; ++i) {
-        ValueJoin::join(functor, shared_ptr,
-                        global_data + word_count.value * i);
+        functor.join(
+            static_cast<pointer_type>(shared_ptr),
+            reinterpret_cast<pointer_type>(global_data + word_count.value * i));
       }
     }
 
-    hip_intra_block_reduce_scan<DoScan, FunctorType, ArgTag>(
-        functor, pointer_type(shared_data));
+    hip_intra_block_reduce_scan<DoScan>(functor, pointer_type(shared_data));
 
     if (DoScan) {
       size_type* const shared_value =
           shared_data +
           word_count.value * (threadIdx.y ? threadIdx.y - 1 : blockDim.y);
 
       if (!threadIdx.y) {
-        ValueInit::init(functor, shared_value);
+        functor.init(reinterpret_cast<pointer_type>(shared_value));
       }
 
       // Join previous inclusive scan value to each member
       for (size_type i = b; i < e; ++i) {
         size_type* const global_value = global_data + word_count.value * i;
-        ValueJoin::join(functor, shared_value, global_value);
-        ValueOps::copy(functor, global_value, shared_value);
+        functor.join(reinterpret_cast<pointer_type>(shared_value),
+                     reinterpret_cast<pointer_type>(global_value));
```

Comment:
Can't you declare `shared_value` and `global_value` to be the correct type instead of all the `reinterpret_cast`?